rfcs/0003-build-contents-diff.md

```diff
@@ -0,0 +1,160 @@
+### Build Contents Diff RFC 
+
+## Context
+In order to make decisions about what to do with any given build generated by TBS, users need to compare a new build’s contents with the contents of other builds.
+
+## Goals
+Create a new command, `kp build diff`, for printing the difference between two builds. The command will take one required argument and two optional arguments.
+ 
+Examples:
+`kp build diff image-name`
+Compares the two most recent builds
+`kp build diff image-name build-number1`
+Compares the stated build and the build minus 1
+`kp build diff image-name build-number1 build-number2`
+Compares the two specified builds
+ 
+Three use cases that this feature should ideally satisfy:
+ 
+- As a DevOps engineer, when I see a new build has been created in a lower environment, I want to evaluate what benefit it provides to my team for me to promote this build to an upper environment. 
+ 
+- As a DevOps engineer or App Developer, when I am told there is a problem with a specific build (new build failing in CI or old build failing in production), I want to know what’s different about the problem build than the last one that worked. 
+
+- As an App Developer, I want to see the difference between the commit I built locally with pack yesterday and the build that kpack built for me today based on my commit. 
+ 
+Implications: 
+- User will need to compare their most recent build to the older build in production
+i.e. builds being compared are non-adjacent 
+- Because the builds are non-adjacent, many things are likely to have changed in the contents and there will have been many “Reasons” that caused the intervening builds. Therefore, it doesn’t make a lot of sense to print the “Reason” as part of the diff
+- I assume that it is easy / possible for users to access the build number associated with a container in production (perhaps there is another build identifier we could use?)
+
+
+
+## Prior Art 
+Differences from the reason diff RFC that’s been added to `kp build status`:
+COMMIT is the same 
+CONFIG is the same
+BUILDPACK is different because it prints the Bill of Materials which represents the results of the buildpacks used instead of only the buildpack/s that caused the new build to kick off
+STACK is the same now, but will be different once we have more stack metadata such as the dpackage list (and someday a CVE list) that we can print
+ 
+Competitor examples: 
+https://github.com/GoogleContainerTools/container-diff
+
+## Complexity/Risks
+- Would this be helpful for diagnosing failed builds or builds that don't start? If so, how can we surface this info without a BoM being generated? 
```

Comment:
It seems that if a build fails the build-reason should sufficiently provide information to diagnose the failure.